src/Polysemy/Internal/Union.hs

```diff
@@ -56,77 +59,79 @@ import Polysemy.Internal.CustomErrors
 ------------------------------------------------------------------------------
 -- | An extensible, type-safe union. The @r@ type parameter is a type-level
 -- list of effects, any one of which may be held within the 'Union'.
-data Union (r :: EffectRow) (m :: Type -> Type) a where
-  Union
-      :: -- A proof that the effect is actually in @r@.
-         ElemOf e r
-         -- The effect to wrap. The functions 'prj' and 'decomp' can help
-         -- retrieve this value later.
-      -> Weaving e m a
-      -> Union r m a
-
-instance Functor (Union r m) where
-  fmap f (Union w t) = Union w $ fmap f t
+data Union (r :: EffectRow) (mWoven :: Type -> Type) a where
+  Union :: !(UnionDetails r mWoven a e) -> Union r mWoven a
+
+data UnionDetails r mWoven a e = UnionDetails
+  -- A proof that the effect is actually in @r@.
+  (ElemOf e r)
+  -- The effect to wrap. The functions 'prj' and 'decomp' can help
+  -- retrieve this value later.
+  (Weaving e mWoven a)
+
+instance Functor (Union r mWoven) where
+  fmap f (Union (UnionDetails w t)) = Union $ UnionDetails w $ fmap f t
   {-# INLINE fmap #-}
 
 
-data Weaving e m a where
+data Weaving e mAfter resultType where
   Weaving
-    :: Functor f
-    =>   { weaveEffect :: e m a
-         -- ^ The original effect GADT originally lifted via
-         -- 'Polysemy.Internal.send'. There is an invariant that @m ~ Sem r0@,
-         -- where @r0@ is the effect row that was in scope when this 'Weaving'
-         -- was originally created.
-       , weaveState :: f ()
-         -- ^ A piece of state that other effects' interpreters have already
-         -- woven through this 'Weaving'. @f@ is a 'Functor', so you can always
-         -- 'fmap' into this thing.
-       , weaveDistrib :: forall x. f (m x) -> n (f x)
-         -- ^ Distribute @f@ by transforming @m@ into @n@. We have invariants
-         -- on @m@ and @n@, which means in actuality this function looks like
-         -- @f ('Polysemy.Sem' (Some ': Effects ': r) x) -> 'Polysemy.Sem' r (f
-         -- x)@.
-       , weaveResult :: f a -> b
-         -- ^ Even though @f a@ is the moral resulting type of 'Weaving', we
-         -- can't expose that fact; such a thing would prevent 'Polysemy.Sem'
-         -- from being a 'Monad'.
-       , weaveInspect :: forall x. f x -> Maybe x
-         -- ^ A function for attempting to see inside an @f@. This is no
-         -- guarantees that such a thing will succeed (for example,
-         -- 'Polysemy.Error.Error' might have 'Polysemy.Error.throw'n.)
-       }
-    -> Weaving e n b
+    :: forall f e rBefore a resultType mAfter. (Functor f)
+    => !(WeavingDetails f e rBefore a resultType mAfter)
+    -> Weaving e mAfter resultType
+
+data Functor f => WeavingDetails f e rBefore a resultType mAfter =
```

Comment:
I agree that it makes sense to remove the Functor constraint from the datatype. I will add another commit to the PR soon to remove this.

I added the `...Details` to make it easier to debug at the type level. The combination of existential quantification and product types makes it hard to constrain the type inference enough to pinpoint why something is wrong. GHC doesn't have any extensions that allow, for example, type applications to constrain the existential type to a scoped type variable.

By splitting the existential part of formulation into the GADT and having a non-existentially quantified Details type, it becomes possible to name and constrain the type variables using existing syntax.

Now that I have got it to compile, it is probably possible to reverse the change to add the Details types; however, given it shouldn't have any performance impact due to it being strict, and given it will likely be helpful to other people trying to change / extend the Polysemy internals, I would suggest leaving it in to improve the developer experience and make it accessible to more people wanting to contribute.

Reply:
I've been hmm-ing and haw-ing about this for a while. I'm conflicted, but ultimately, I'm leaning towards "no". `-Details` clog up the machinery, and to gain access to existentially quantified type variables, there are easy targets both in `Weaving` and in `Union`: in `Weaving`, you can pattern match on the functorial state to gain access to the functor, and the effect to gain access to the existentially quantified row. On `Union`, you can gain access to the existentially quantified effect by pattern matching on `Weaving` or `ElemOf`. I agree, as it is now it poses a barrier to entry, but not, I think, a very big one (compared to the other stuff you need to learn for working with polysemy internals); and besides, I don't expect many newcomers to tinker with the internals.

Reply:
I have added another commit to remove the `-Details` layer.